crates/astria-sequencer/src/transaction/fees.rs

```diff
@@ -0,0 +1,683 @@
+        .await
+        .wrap_err("failed to get transfer base fee")?;
+    let sequence_base_fee = state
+        .get_sequence_action_base_fee()
+        .await
+        .wrap_err("failed to get base fee")?;
+    let sequence_byte_cost_multiplier = state
+        .get_sequence_action_byte_cost_multiplier()
+        .await
+        .wrap_err("failed to get fee per byte")?;
+    let init_bridge_account_base_fee = state
+        .get_init_bridge_account_base_fee()
+        .await
+        .wrap_err("failed to get init bridge account base fee")?;
+    let bridge_lock_byte_cost_multiplier = state
+        .get_bridge_lock_byte_cost_multiplier()
+        .await
+        .wrap_err("failed to get bridge lock byte cost multiplier")?;
+    let bridge_sudo_change_fee = state
+        .get_bridge_sudo_change_base_fee()
+        .await
+        .wrap_err("failed to get bridge sudo change fee")?;
+    let ics20_withdrawal_base_fee = state
+        .get_ics20_withdrawal_base_fee()
+        .await
+        .wrap_err("failed to get ics20 withdrawal base fee")?;
+    Ok(Fees {
+        transfer_base_fee,
+        sequence_base_fee,
+        sequence_byte_cost_multiplier,
+        init_bridge_account_base_fee,
+        bridge_lock_byte_cost_multiplier,
+        bridge_sudo_change_fee,
+        ics20_withdrawal_base_fee,
+    })
+}
+
+/// Adds the fee amount to be paid by the from to the to along with an event of kind
+/// "tx.fees" to the payment map
+fn increase_fees(
+    from: [u8; 20],
+    to: [u8; 20],
+    fee_asset: &asset::Denom,
+    amount: u128,
+    fee_payment_map: &mut HashMap<([u8; 20], [u8; 20], asset::Denom), FeeInfo>,
+    action_type: String,
+    action_index: u32,
+) {
+    let fee_event = construct_tx_fee_event(fee_asset, amount, action_type, action_index);
+    fee_payment_map
+        .entry((from, to, fee_asset.clone()))
+        .and_modify(|fee_info| {
+            fee_info.amt = fee_info.amt.saturating_add(amount);
+            fee_info.events.push(fee_event.clone());
+        })
+        .or_insert(FeeInfo {
+            amt: amount,
+            events: vec![fee_event],
+        });
+}
+
+/// Creates `abci::Event` of kind `tx.fees` for sequencer fee reporting
+pub(crate) fn construct_tx_fee_event<T: std::fmt::Display>(
+    asset: &T,
+    fee_amount: u128,
+    action_type: String,
+    action_index: u32,
+) -> Event {
+    Event::new(
+        "tx.fees",
+        [
+            ("asset", asset.to_string()).index(),
+            ("feeAmount", fee_amount.to_string()).index(),
+            ("actionType", action_type).index(),
+            ("actionIndex", action_index.to_string()).index(),
+        ],
+    )
+}
+
+#[instrument(skip_all)]
+fn warn_if_asset_type_not_allowed(asset: &asset::Denom, allowed_fee_assets: &[IbcPrefixed]) {
+    if !allowed_fee_assets.contains(&IbcPrefixed::from(asset)) {
+        warn!("Transaction execution will fail, asset type not allowed for fee payment: {asset}");
+    }
+}
```

Comment:
this should be a hard error, not just a log

Reply:
Should this be a hard error even if the transaction is not being executed? This function may be called by `transaction_fee_request`, in which case I had thought it might be important to differentiate between warning when `get_and_report_tx_fees()` is called for a fee request, versus erroring when it is called during transaction execution.